madara/crates/client/rpc/src/versions/user/v0_7_1/methods/read/syncing.rs

```diff
@@ -15,32 +18,46 @@ use crate::Starknet;
 ///   sync status, or a `Boolean` (`false`) indicating that the node is not currently synchronizing.
+///
+/// Following the spec: https://github.com/starkware-libs/starknet-specs/blob/2030a650be4e40cfa34d5051a0334f375384a421/api/starknet_api_openrpc.json#L765
+/// if the node is synced it will return a SyncingStatus::NotSyncing which is a boolean false, and in case of syncing it will return a SyncStatus struct
 pub async fn syncing(starknet: &Starknet) -> StarknetRpcResult<SyncingStatus> {
-    // obtain best seen (highest) block number
-    let Some(current_block_info) = starknet
+    // Get current block info first
+    let (current_block_num, current_block_hash) = match starknet
         .backend
         .get_block_info(&BlockId::Tag(BlockTag::Latest))
         .or_internal_server_error("Error getting latest block")?
-    else {
-        return Ok(SyncingStatus::NotSyncing); // TODO: This doesn't really make sense? This can only happen when there are no block in the db at all.
+        .ok_or_internal_server_error("Latest block not found")
+    {
+        Ok(block_info) => {
+            let current_block_info =
+                block_info.as_nonpending().ok_or_internal_server_error("Latest block cannot be pending")?;
+            let current_block_num = current_block_info.header.block_number;
+            let current_block_hash = current_block_info.block_hash;
+            (current_block_num, current_block_hash)
+        }
+        Err(_err) => (0u64, Felt::ZERO),
     };
 
-    let current_block_info =
-        current_block_info.as_nonpending().ok_or_internal_server_error("Latest block cannot be pending")?;
-    let starting_block_num = 0; // TODO(rpc): fix this // starknet.starting_block;
-    let starting_block_info = starknet.get_block_info(&BlockId::Number(starting_block_num))?;
-    let starting_block_info =
-        starting_block_info.as_nonpending().ok_or_internal_server_error("Block cannot be pending")?;
-    let starting_block_hash = starting_block_info.block_hash;
-    let current_block_num = current_block_info.header.block_number;
-    let current_block_hash = current_block_info.block_hash;
+    // Get the sync status from starknet which in turn gets it from MadaraBackend
+    let sync_status =
+        starknet.sync_status().await.or_internal_server_error("Error getting sync status after retries")?;
 
-    Ok(SyncingStatus::Syncing(SyncStatus {
-        starting_block_num,
-        starting_block_hash,
-        highest_block_num: current_block_num, // TODO(merge): is this correct?
-        highest_block_hash: current_block_hash,
-        current_block_num,
-        current_block_hash,
-    }))
+    match sync_status {
+        StarknetSyncStatus::NotRunning => Ok(SyncingStatus::NotSyncing),
+        StarknetSyncStatus::Running { starting_block_n, starting_block_hash, highest_block_n, highest_block_hash } => {
+            if highest_block_n - SYNC_THRESHOLD_BLOCKS <= current_block_num {
```

Comment:
this will panic in debug if highest_block_n is <= 6
use
```rs
if current_block_num + SYNC_THRESHOLD_BLOCKS >= highest_block_n {
```

Reply:
oh, your condition is reversed
you meant `>=` not `<=`, so replace it with
```rs
if current_block_num + SYNC_THRESHOLD_BLOCKS <= highest_block_n {
```
it would really help if there was an integration test for the endpoint but i would prefer to merge the pr asap so don't bother